src/main/java/com/zendesk/maxwell/schema/MysqlSchemaCompactor.java

```diff
@@ -0,0 +1,197 @@
+		if ( !shouldCompact(cx) ) {
+			return null;
+		}
+
+		ResultSet rs = cx.prepareStatement(
+			"select id, binlog_file, binlog_position, gtid_set, 0 as last_heartbeat_read "
+			+ " from `schemas` where `server_id` = " + this.serverID
+			+ " order by id desc limit 1"
+		).executeQuery();
+
+		if ( !rs.next() )
+			return null;
+
+
+		Long schemaID = rs.getLong("id");
+		Position schemaPosition = MysqlPositionStore.positionFromResultSet(rs, serverID == 0);
+
+		LOGGER.debug("trying to compact schema {} @ {}", schemaID, schemaPosition);
+
+		ResultSet positionsRS = cx.prepareStatement(
+			"select * from `positions` where server_id = " + serverID
+		).executeQuery();
+
+		while ( positionsRS.next() ) {
+			Position clientPosition = MysqlPositionStore.positionFromResultSet(positionsRS, serverID == 0);
+			if ( clientPosition.newerThan(schemaPosition) ) {
+				LOGGER.debug("found a client @ {}, that's fine...", clientPosition);
+			} else {
+				if ( !schemaID.equals(lastWarnedSchemaID) ) {
+					LOGGER.warn("Not compacting schema {}, client '{}' @ {} has not reached that position yet",
+						schemaID,
+						positionsRS.getString("client_id"),
+						clientPosition);
+					lastWarnedSchemaID = schemaID;
+				}
+				return null;
+			}
+		}
+
+		return schemaID;
+	}
+
+	private void compact(Connection cx) throws SQLException, InvalidSchemaError {
+
+		if ( shouldCompact(cx) ) {
+
+			Long schemaID = chooseCompactedSchemaBase(cx);
+			if ( schemaID == null)
+				return;
+
+			LOGGER.info("compacting schemas before {}", schemaID);
+			cx.prepareStatement("BEGIN").execute();
+
+			MysqlSavedSchema savedSchema = MysqlSavedSchema.restoreFromSchemaID(schemaID, cx, this.sensitivity);
+			savedSchema.saveFullSchema(cx, schemaID);
+			cx.createStatement().executeUpdate("update `schemas` set `base_schema_id` = null, `deltas` = null where `id` = " + schemaID);
+
+			cx.prepareStatement("COMMIT").execute();
+
+			slowDeleteSchemas(cx, schemaID);
+		}
+	}
+
+	private void slowDeleteSchemas(Connection cx, long newBaseSchemaID) throws SQLException {
+		cx.setAutoCommit(true);
+
+		PreparedStatement ps = cx.prepareStatement(
+			"select * from `schemas` where id < ? and server_id = ?"
+		);
+		ps.setLong(1, newBaseSchemaID);
+		ps.setLong(2, serverID);
+
+		ResultSet rs = ps.executeQuery();
+		while ( rs.next() ) {
+			slowDeleteSchema(cx, rs.getLong("id"));
+		}
+	}
+
+	private void slowDeleteSchema(Connection cx, long schemaID) throws SQLException {
+		LOGGER.debug("slow deleting schema_id: {}", schemaID);
+		slowDeleteFrom("columns", cx, schemaID);
+		slowDeleteFrom("tables", cx, schemaID);
+		slowDeleteFrom("databases", cx, schemaID);
+		cx.createStatement().executeUpdate("delete from `schemas` where id = " + schemaID);
+	}
+
+	private static final int DELETE_SLEEP_MS = 200;
+	private static final int DELETE_LIMIT = 500;
+
+	private void slowDeleteFrom(String table, Connection cx, long schemaID) throws SQLException {
```

Comment:
Is this just to prevent a huge transaction? And I know these strings are constant-ish, but can you use a prepared statement for the table name + ID?

Reply:
yeah mostly to prevent a huge transaction.  Mysql also does some weird shit with like compaction threads when you delete a ton of rows at once (although that may be a concern from a different time, don't remember).

I don't think I can inject a table name into a prepared statement.  IDs I can, but other than a marginal performance benefit I'm not sure of the value.